### PR TITLE
Add an Arduino example with documentation.

### PR DESCRIPTION
**Pull request by haydenroche5**

### CODE REVIEW

File: src/NotecardEnvVarManager.h [hunk 1]
```diff
@@ -7,7 +7,6 @@ extern "C" {
 #endif
 
 enum {
-    NEVM_WAITING = -2,
```
Comment (haydenroche5) [inline src/NotecardEnvVarManager.h:10]: Not needed after removing the latent cache stuff.

File: examples/arduino/arduino.ino [hunk 2]
```diff
@@ -0,0 +1,116 @@
+#include <Wire.h>
```
Comment (zfields) [inline examples/arduino/arduino.ino:1]: It is not necessary to include `Wire.h`.
```suggestion
```

File: examples/arduino/arduino.ino [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+void envVarManagerCb(const char *var, const char *val, void *userCtx)
+{
+    EnvVarCache *cache = (EnvVarCache *)userCtx;
+
+    // Cache the values for each variable.
+    if (strcmp(var, "variable_a") == 0) {
+        strncpy(cache->valueA, val, sizeof(cache->valueA));
+    }
+    else if (strcmp(var, "variable_b") == 0) {
+        strncpy(cache->valueB, val, sizeof(cache->valueA));
```
Comment (zfields) [inline examples/arduino/arduino.ino:41]:
```suggestion
        strncpy(cache->valueB, val, sizeof(cache->valueB));
```

File: examples/arduino/arduino.ino [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    }
+    else if (strcmp(var, "variable_c") == 0) {
+        strncpy(cache->valueC, val, sizeof(cache->valueA));
```
Comment (zfields) [inline examples/arduino/arduino.ino:44]:
```suggestion
        strncpy(cache->valueC, val, sizeof(cache->valueC));
```

File: examples/arduino/arduino.ino [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+void setup()
+{
+    delay(2500);
+    Serial.begin(115200);
+    // Set up debug output via serial connection.
+    notecard.setDebugOutputStream(Serial);
+
+    // Initialize the physical I/O channel to the Notecard.
+    Wire.begin();
```
Comment (zfields) [inline examples/arduino/arduino.ino:64]: Unnecessary to call Wire.begin();
```suggestion
```

File: examples/arduino/arduino.ino [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    "variable_b",
+    "variable_c"
+};
+static const size_t numEnvVars = sizeof(envVars) / sizeof(envVars[0]);
+
+void setup()
+{
+    delay(2500);
+    Serial.begin(115200);
+    // Set up debug output via serial connection.
```
Comment (zfields) [inline examples/arduino/arduino.ino:60]: This is how we handle awaiting `Serial` in the other examples...

```suggestion
    // If you open Arduino's serial terminal window, you'll be able to watch
    // JSON objects being transferred to and from the Notecard for each request.
    Serial.begin(115200);
    const size_t serial_timeout_ms = 3000;
    for (const size_t start_ms = millis(); !Serial && (millis() - start_ms) < serial_timeout_ms;)
        ;
```

File: examples/arduino/arduino.ino [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        &envVarCache) != NEVM_SUCCESS) {
+        notecard.logDebug("Failed to set env var manager callback.\r\n");
+        return;
+    }
+}
+
+void loop()
+{
+    // Fetch the environment variables every FETCH_INTERVAL_MS milliseconds.
+    uint32_t currentMs = NoteGetMs();
```
Comment (zfields) [inline examples/arduino/arduino.ino:102]: `NoteGetMs()` is an abstraction that we typically use internally for libraries that may not be written to the Arduino API. Personally, I would use `millis()` here, but this is harmless.

```suggestion
    uint32_t currentMs = millis();
```

File: examples/arduino/arduino.ino [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    if (!notecard.sendRequestWithRetry(req, HUB_SET_RETRY_SECONDS)) {
+        notecard.logDebug("hub.set request failed.\r\n");
+        return;
+    }
+
+    // Allocate the environment variable manager.
+    envVarManager = NotecardEnvVarManager_alloc();
+    if (envVarManager == NULL) {
+        notecard.logDebug("Failed to allocate env var manager.\r\n");
+        return;
```
Comment (zfields) [inline examples/arduino/arduino.ino:86]: The example has failed at this point. If you `return` you move onto `loop()`.

```suggestion
        while(1);
```

File: examples/arduino/arduino.ino [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+       JAddStringToObject(req, "product", PRODUCT_UID);
+    }
+    JAddStringToObject(req, "mode", "continuous");
+    JAddBoolToObject(req, "sync", true);
+    // Send the request with a retry timeout. If the Notecard has just started
+    // up, it may need a moment before it's able to receive and respond to
+    // requests.
+    if (!notecard.sendRequestWithRetry(req, HUB_SET_RETRY_SECONDS)) {
+        notecard.logDebug("hub.set request failed.\r\n");
+        return;
```
Comment (zfields) [inline examples/arduino/arduino.ino:79]: The example has failed at this point. If you `return` you move onto `loop()`.

```suggestion
        while(1);
```

File: examples/arduino/arduino.ino [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    // Set the callback for the manager, and give it a pointer to our cache
+    // so that we can store the values of the environment variables we're
+    // fetching.
+    if (NotecardEnvVarManager_setEnvVarCb(envVarManager, envVarManagerCb,
+        &envVarCache) != NEVM_SUCCESS) {
+        notecard.logDebug("Failed to set env var manager callback.\r\n");
+        return;
```
Comment (zfields) [inline examples/arduino/arduino.ino:95]: The example has failed at this point. If you `return` you move onto `loop()`.

```suggestion
        while(1);
```

**Review by zfields — CHANGES_REQUESTED** [review]
There is some [other boilerplate](https://github.com/blues/note-arduino/blob/master/examples/Example5_UsingTemplates/Example5_UsingTemplates.ino#L34) we typically add to let folks choose between TX/RX and I2C, but I don't think it's strictly necessary.

In my comments, I updated this with the minimum boilerplate that fits what you have already put forth.

**Comment by zfields** [comment]
[Add an Arduino example for the NotecardEnvVarManager](https://trello.com/c/7d2eXyvm/102-add-an-arduino-example-for-the-notecardenvvarmanager)

**Review by zfields — APPROVED** [review]

**Pull request merged**